src/phoenix/core/traces.py

```diff
@@ -1,108 +1,267 @@
+        super().__init__(span)
+        self._self_computed_values: Dict[str, SupportsFloat] = {}
+
+    @property
+    def span(self) -> Span:
+        span = decode(self.__wrapped__)
+        span.attributes.update(cast(SpanAttributes, self._self_computed_values))
+        return span
+
+    def __getitem__(self, key: str) -> Any:
+        if key.startswith(COMPUTED_PREFIX):
+            return self._self_computed_values.get(key)
+        if key.startswith(CONTEXT_PREFIX):
+            suffix_key = key[len(CONTEXT_PREFIX) :]
+            return getattr(self.__wrapped__.context, suffix_key, None)
+        if key.startswith(ATTRIBUTE_PREFIX):
+            suffix_key = key[len(ATTRIBUTE_PREFIX) :]
+            if suffix_key not in self.__wrapped__.attributes:
+                return None
+            return self.__wrapped__.attributes[suffix_key]
+        return getattr(self.__wrapped__, key, None)
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        if not key.startswith(COMPUTED_PREFIX):
+            raise KeyError(f"{key} is not a computed value")
+        self._self_computed_values[key] = value
+
+
+ParentSpanID: TypeAlias = SpanID
+ChildSpanID: TypeAlias = SpanID
 
 
 class Traces:
-    """
-    Traces class is used to contain abstractions around the traces dataframe
-    """
-
-    def __init__(self, df: pd.DataFrame):
-        df = df.set_index(SPAN_ID, drop=False)
-        df[LATENCY_MS] = (df[END_TIME] - df[START_TIME]).dt.total_seconds() * 1000
-        parent_span_ids = df.loc[:, PARENT_ID]
-        for column, cumulative_column in (
+    def __init__(self, spans: Optional[Iterable[Span]] = None) -> None:
+        self._queue: "SimpleQueue[Optional[pb.Span]]" = SimpleQueue()
+        # Putting `None` as the sentinel value for queue termination.
+        weakref.finalize(self, self._queue.put, None)
+        for span in spans or ():
+            self.put(span)
+        self._lock = RLock()
+        self._spans: Dict[SpanID, ReadableSpan] = {}
+        self._parent_span_ids: Dict[SpanID, ParentSpanID] = {}
+        self._traces: Dict[TraceID, List[SpanID]] = defaultdict(list)
+        self._child_span_ids: DefaultDict[SpanID, List[ChildSpanID]] = defaultdict(list)
+        self._orphan_spans: DefaultDict[ParentSpanID, List[pb.Span]] = defaultdict(list)
+        self._start_time_sorted_span_ids: SortedKeyList[SpanID] = SortedKeyList(
+            key=lambda span_id: self._spans[span_id].start_time.ToDatetime(timezone.utc),
+        )
+        self._start_time_sorted_root_span_ids: SortedKeyList[SpanID] = SortedKeyList(
+            key=lambda span_id: self._spans[span_id].start_time.ToDatetime(timezone.utc),
+        )
+        self._latency_sorted_root_span_ids: SortedKeyList[SpanID] = SortedKeyList(
+            key=lambda span_id: self._spans[span_id][LATENCY_MS],
+        )
+        self._min_start_time: Optional[datetime] = None
+        self._max_start_time: Optional[datetime] = None
+        self._start_consumer()
```

Comment:
We can leave it for now but I do still have this concern where transport code ends up proliferating business logic. It's something that can cause a cascade of strange dependancies.

I agree the encapsulation of a stream makes sense but there are benefits to having derived streams that are observable on their own merit. 

Let's see how it goes.

Reply:
ok. so it seems to me that having it decoded at te receiver (i.e. at the `v1/spans` endpoint) would resolve your concern because Traces now no longer knows about protos or anything from the transportation layer